Replace HTTP replication with TCP replication (Server side part)

erikjohnston opened the pull request:
This is the server side component of https://github.com/matrix-org/synapse/pull/2069, including the requested changes. 

Docs [rendered](https://github.com/matrix-org/synapse/blob/erikj/repl_tcp_server/docs/tcp_replication.rst)

erikjohnston commented on synapse/notifier.py:
```diff
@@ -550,3 +550,9 @@ def wait_for_replication(self, callback, timeout):
+        """Returns a deferred which resolves when there is new data for
+        replication to handle.
+        """
+        return self.replication_deferred.observe()
```
It may be nicer to let the replication resource register a callback rather than bouncing through a deferred?

richvdh replied:
yes, I think it would. it may be more efficient too.

richvdh replied:
otherwise I think it needs a `make_deferred_yieldable`.

erikjohnston commented on synapse/replication/tcp/resource.py:
```diff
@@ -0,0 +1,300 @@
+        connections if there are.
+
+        This should get called each time new data is available, even if it
+        is currently being executed, so that nothing gets missed
```
Maybe it would be a lot clearer if this function looped until there were no changes? Although that is less immediately obvious is correct.

It may be more obvious if we can register this function directly with the notifer, rather than bouncing via a deferred in a loop.

richvdh replied:
+1 to registering a callback
-1 to looping until there are no changes. It will complicate this function and I don't think it will make anything clearer.

richvdh commented on synapse/handlers/presence.py:
```diff
@@ -509,6 +511,71 @@ def update_external_syncs(self, process_id, syncing_user_ids):
+                updates.append(prev_state.copy_and_replace(
+                    last_user_sync_ts=time_now_ms,
+                ))
+                process_presence.discard(user_id)
```
we'll hit this if `is_syncing and user_id in process_presence`. I think it's incorrect.

richvdh commented on synapse/handlers/presence.py:
```diff
@@ -509,6 +511,71 @@ def update_external_syncs(self, process_id, syncing_user_ids):
+            if updates:
+                yield self._update_states(updates)
+
+            self.external_process_last_updated_ms[process_id] = self.clock.time_msec()
```
just use time_now_ms ?

richvdh commented on synapse/replication/tcp/__init__.py:
```diff
@@ -0,0 +1,46 @@
+ * resource.py - the server classes that accepts and handle client connections
+ * streams.py  - the definitons of all the valid streams
+
+Further details can be found in docs/tcp_replication.rst
```
suggest moving this up a bit to where you describe the protocol. I nearly missed it here.

Tbh I'm not sure you actually need the description of the protocol in this file - just referring to the doc would be fine imho

richvdh commented on synapse/replication/tcp/__init__.py:
```diff
@@ -0,0 +1,46 @@
+    > RDATA events 55 ["$foo4:bar.com", ...]
+
+The example shows the server accepting a new connection and sending its identity
+with the `SERVER` command, followed by the client asking to subscribe to the
```
I'm not going to insist you go through and change them, but for future reference: AIUI docstrings are interpreted as RST and really ought to have double-backticks.

richvdh commented on synapse/storage/pusher.py:
```diff
@@ -135,6 +135,48 @@ def get_all_updated_pushers_txn(txn):
+        """Get all the pushers that have changed between the given tokens.
+
+        Returns:
+            list(tuple): each tuple consists of:
```
No, it returns a Deferred, which resolves to that lot.

(I'd prefer it if we even put this on `@defer.inlineCallbacks`, but I won't insist there because it's "obvious". Here though, there is no clue that it actually returns a Deferred, rather than a list.)

richvdh commented on synapse/replication/tcp/streams.py:
```diff
@@ -0,0 +1,409 @@
+        True then limit is provided, otherwise it's not.
+
+        Returns:
+            list(tuple): the first entry in the tuple is the token for that
```
If it is valid for this function to return a Deferred, say so.

erikjohnston commented on docs/tcp_replication.rst:
```diff
@@ -0,0 +1,177 @@
+
+Blank lines are ignored.
+
+
```
Done.

erikjohnston commented:
I think I've addressed all the issues raised.

I also cheekily added a `<time_msec>` to the `USER_SYNC` command (which is the last sync time), as I think a) its more correct and b) is required if we want to batch up the USER_SYNC stuff on the workers to reduce traffic.
